File: sample_apps/goV2/cleanup-sample.go
package main

import (
	"context"
	"errors"
	"flag"
	"fmt"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/timestreamquery"
	"github.com/aws/aws-sdk-go-v2/service/timestreamwrite"
	"github.com/aws/aws-sdk-go-v2/service/timestreamwrite/types"
	"go_sample/utils"
	"net/http"
)

func main() {
	// process command line arguments
	region := flag.String("region", "us-east-1", "region")
	flag.Parse()

	databaseName := utils.DATABASE_NAME
	tableName := utils.TABLE_NAME

	tr := utils.LoadHttpSettings()
	// Use the SDK's default configuration.
	cfg, err := config.LoadDefaultConfig(context.TODO())
	cfg.Region = *region
	cfg.HTTPClient = &http.Client{Transport: tr}

	utils.HandleError(err, "Failed to load config ", true)

	writeSvc := timestreamwrite.NewFromConfig(cfg)
	querySvc := timestreamquery.NewFromConfig(cfg)

	s3Svc := s3.NewFromConfig(cfg)

	timestreamBuilder := utils.TimestreamBuilder{WriteSvc: writeSvc, QuerySvc: querySvc}
	timestreamDependencyHelper := utils.TimestreamDependencyHelper{S3Svc: s3Svc}
	var RESOURCE_NOT_FOUND *types.ResourceNotFoundException

	fmt.Printf("Describing a table with name %s\n", tableName)
	// Describe table.
	describeTableOutput, err := timestreamBuilder.DescribeTable(databaseName, tableName)
	if err != nil {
		if errors.As(err, &RESOURCE_NOT_FOUND) {
			fmt.Printf("Table %s was already deleted/does not exist\n\n", tableName)
		} else {
			utils.HandleError(err, fmt.Sprintf("Failed to describe table='%s'\n", tableName), true)
		}
	} else {
		if *describeTableOutput.Table.MagneticStoreWriteProperties.EnableMagneticStoreWrites == true &&
			describeTableOutput.Table.MagneticStoreWriteProperties.MagneticStoreRejectedDataLocation != nil &&
			describeTableOutput.Table.MagneticStoreWriteProperties.MagneticStoreRejectedDataLocation.S3Configuration != nil {
			s3BucketName := *describeTableOutput.Table.MagneticStoreWriteProperties.MagneticStoreRejectedDataLocation.S3Configuration.BucketName
			fmt.Printf("Deleting S3 bucket associated with the table. s3BucketName='%s'\n", s3BucketName)
			err = timestreamDependencyHelper.DeleteS3Bucket(s3BucketName)
			if err != nil {
				fmt.Printf("Failed to delete s3BucketName='%s', please delete it manually\n", s3BucketName)
			}
		}

		fmt.Printf("Deleting table='%s'\n", tableName)
		err = timestreamBuilder.DeleteTable(databaseName, tableName)
		utils.HandleError(err, fmt.Sprintf("Failed to delete table='%s'\n\n", tableName), true)
	}

	fmt.Printf("Deleting database='%s'\n", databaseName)
	err = timestreamBuilder.DeleteDatabase(databaseName)
	if err != nil && errors.As(err, &RESOURCE_NOT_FOUND) {
		fmt.Printf("Database %s was already deleted/does not exist\n", databaseName)
	}
}

File: sample_apps/rust/examples/cleanup_sample.rs
use anyhow::{anyhow, Error, Result};
use aws_sdk_timestreamwrite as timestream_write;
use aws_sdk_timestreamwrite::operation::describe_table::DescribeTableOutput;
use clap::Parser;
use sample_timestream_app::utils::timestream_helper;

#[tokio::main]
async fn main() -> Result<(), Error> {
    let args = timestream_helper::Args::parse();

    let client = timestream_helper::get_connection(&args.region)
        .await
        .expect("Failed to get connection to Timestream");

    match client
        .describe_table()
        .database_name(&args.database_name)
        .table_name(&args.table_name)
        .send()
        .await
    {
        Ok(describe_table_output) => {
            if let Some(s3_config) = get_magnetic_storage_configuration(&describe_table_output) {
                let bucket_name = s3_config
                    .bucket_name()
                    .expect("Failed to retrieve bucket name");
                match timestream_helper::delete_s3_bucket(bucket_name, &args.region).await {
                    Ok(_) => {
                        println!("Successfully deleted s3 bucket {:?} ", bucket_name)
                    }
                    Err(err) => println!(
                        "Failed to delete s3 bucket {:?}, err: {:?}",
                        bucket_name, err
                    ),
                }
            }

            client
                .delete_table()
                .database_name(&args.database_name)
                .table_name(&args.table_name)
                .send()
                .await?;
            println!("Successfully deleted table {:?}", args.table_name);
        }

        Err(describe_table_error) => {
            if describe_table_error
                .as_service_error()
                .map(|e| e.is_resource_not_found_exception())
                == Some(true)
            {
                println!(
                    "Skipping table deletion as the table {:?} does not exist",
                    args.table_name
                );
            } else {
                return Err(anyhow!(
                    "Failed to describe the table {:?}, Error: {:?}",
                    args.table_name,
                    describe_table_error
                ));
            }
        }
    }

    match client
        .describe_database()
        .database_name(&args.database_name)
        .send()
        .await
    {
        Ok(_) => match client
            .delete_database()
            .database_name(&args.database_name)
            .send()
            .await
        {
            Ok(_) => println!("Successfully deleted database {:?}", &args.database_name),
            Err(err) => println!(
                "Failed to delete database {:?}, err: {:?}",
                &args.database_name, err
            ),
        },
        Err(describe_database_error) => {
            if describe_database_error
                .as_service_error()
                .map(|e| e.is_resource_not_found_exception())
                == Some(true)
            {
                println!(
                    "Skipping database deletion as the database {:?} does not exist",
                    args.database_name
                );
            } else {
                return Err(anyhow!(
                    "Failed to describe the database {:?}, Error: {:?}",
                    args.database_name,
                    describe_database_error
                ));
            }
        }
    }

    Ok(())
}

fn get_magnetic_storage_configuration(
    describe_table_output: &DescribeTableOutput,
) -> Option<&timestream_write::types::S3Configuration> {
    let table = describe_table_output.table()?;
    let magnetic_store_write_properties = table.magnetic_store_write_properties()?;

    if magnetic_store_write_properties.enable_magnetic_store_writes() {
        let rejected_data_location =
            magnetic_store_write_properties.magnetic_store_rejected_data_location()?;
        return rejected_data_location.s3_configuration();
    }
    None
}
